### PR TITLE
47 hardware implementation sdi12

### LINKED ISSUE #47: Hardware implementation of SDI-12 (Due: 2024-03-12)
The current SDI-12 implementation uses software SDI-12 with NOPS which has been shown to be unreliable. A hardware implementation would reduce the possibility of timing errors resulting from other code running on the device. We can also utilize the LPUART interface to reduce power consumption during long communication times.

Outcome: Replicate results from software SDI-12

Resources:
This link shows how to implement hardware SDI-12 with a few components and gives CubeMX code examples: https://permalink.lanl.gov/object/tr?what=info:lanl-repo/lareport/LA-UR-15-26961. A copy also lives in our private gdrive.

### PR DESCRIPTION
**Pull request by stevegtaylor**
---
name: SDI-12 Pull Request
about: Review some design decisions in SDI-12 library
assignees: @stevegtaylor
reviewers: @jmadden173
---

**General Design Review of SDI-12 before further integration**
#47 
I would like to review the SDI-12 before the final implantation.

Closes #47

### CODE REVIEW

File: stm32/Src/i2c.c
Comment (jmadden173) [inline stm32/Src/i2c.c:1]: Why was this changed?

File: stm32/Src/main.c [hunk 1]
```diff
@@ -152,8 +167,14 @@ void SystemClock_Config(void)
   RCC_OscInitStruct.LSEState = RCC_LSE_ON;
   RCC_OscInitStruct.MSIState = RCC_MSI_ON;
   RCC_OscInitStruct.MSICalibrationValue = RCC_MSICALIBRATION_DEFAULT;
-  RCC_OscInitStruct.MSIClockRange = RCC_MSIRANGE_11;
-  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+  RCC_OscInitStruct.MSIClockRange = RCC_MSIRANGE_6;
```
Comment (jmadden173) [inline stm32/Src/main.c:170]: Why is the RTC clock changed? This may cause issues with the LoRa libraries since it requires a specific clock configuration so that it will run while the chip is in STOP2 mode. Have you checked it still works?

File: stm32/Src/sdi12_v1.c
Comment (jmadden173) [inline stm32/Src/sdi12_v1.c:1]: Same as header, delete

File: stm32/Src/sdi12.c [hunk 2]
```diff
@@ -3,54 +3,108 @@
...
 
-    // Configure GPIO pin as input
-    GPIO_InitStruct.Pin = GPIO_PIN_2;
-    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
-    GPIO_InitStruct.Pull = GPIO_NOPULL;
-    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+HAL_StatusTypeDef SDI12_SendCommand(const char *command, uint8_t size){
+    HAL_StatusTypeDef ret;
+    SDI12_WakeSensors();
+    ret = HAL_UART_Transmit(&huart2, (const uint8_t *) command, size, 1000);
```
Comment (jmadden173) [inline stm32/Src/sdi12.c:91]: Make the timeout some static const so that its shared between all commands and can be configured easily

File: stm32/platformio.ini [hunk 3]
```diff
@@ -36,25 +36,43 @@ build_src_filter = +<*> -<.git/> -<examples/*>
 # stlink
 debug_tool = stlink
 upload_protocol = stlink
-#
-#extra_scripts = pre:tool_openocd.py  ; Add this line
-#
+
+extra_scripts = pre:tool_openocd.py  ; Add this line
```
Comment (jmadden173) [inline stm32/platformio.ini:40]: Duplicate lines

File: stm32/platformio.ini [hunk 4]
```diff
@@ -36,25 +36,43 @@ build_src_filter = +<*> -<.git/> -<examples/*>
...
 monitor_speed = 115200
 
-test_port = /dev/ttyUSB0
+test_port = COM12
 test_speed = 115200
 
-#debug_init_break = tbreak main
-
-#board_build.stm32cube.custom_config_header = yes
+[env_debug]
```
Comment (jmadden173) [inline stm32/platformio.ini:59]: Why is this section still here? I removed it when I fixed debugging

File: stm32/stm32pio.ini
Comment (jmadden173) [inline stm32/stm32pio.ini:1]: Revert file to keep paths for both windows and linux

**Review by jmadden173 — CHANGES_REQUESTED** [review]
I think in the future we should limit merging between development branches. I see a lot of my code here and it makes it somewhat hard to go through. An example produce for this case would be (1) I create PR with incremental change (2) someone review/approves and it merged into main (3) You can pull the update main branch. From this POV it looks like you changed the FIFO and stuff.

From the [style guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros) we should be using `const` instead of `#defines`. I think that defines are still relevant considering thats one of the few ways to change options outside of the codebase. But they still should be set to an actual variables before use to prevent type related errors.

For issue #36, it would be helpful to have the public interface to the library contained in the library itself. Aka I think we should be defining our own error messages rather than using `HAL_Status_Typedef`. It would allow for finer grain of errors rather than `HAL_OK` or `HAL_ERROR`. This was super helpful when I was writing the fram code, I would check that out if its still unclear.

In the docstrings its unclear what timer is currently used for the SDI-12 library. Multiple timers are configured but it would require someone to find the code to see which one. I think that the TEROS-12 specific code should be moved into a separate library with SDI-12 as a dependency. It would make #51 a lot easier since you wouldn't have to modify the SDI-12 code and re-test teros12.

There should be an automatic unit test for SDI-12 that checks he interfaces functionality. It will make unit testing much quicker. Similar to the PR for smoke tests for ADC and battery in #56 

Any lines that are over 80 width should be update based on the style guide. See [here](https://google.github.io/styleguide/cppguide.html#Line_Length)

**Comment by jmadden173** [comment]
In future PRs can you add more description to the implementation details and follow the PR template where applicable.

**Comment by stevegtaylor** [comment]
added changes see commit d5bad87961af6ebd8a132fe540d2acf1a0a41707

Still having initialization issues with the RTC, I'm hoping a merge with @jmadden173's #67 will resolve the initialization issues.

**Comment by jmadden173** [comment]
Hmm, I'm wondering whats causing these RTC issues. This seems to be a recurring problem.

John

File: stm32/.gitignore [hunk 5]
```diff
@@ -3,6 +3,9 @@
...
 .vscode/launch.json
 .vscode/ipch
+.vscode/settings.json
 
 # docs
-docs/html
\ No newline at end of file
+docs/html
+
+stm32pio.ini
```
Comment (jmadden173) [inline stm32/.gitignore:11]: I like this, I think we just make a template and turn the file into a per user configuration. We should provide a template file and `stm32pio.ini` should be removed from git.

File: stm32/test/test_sdi12/test_sdi12.c [hunk 6]
```diff
@@ -0,0 +1,164 @@
+/**
```
Comment (jmadden173) [inline stm32/test/test_sdi12/test_sdi12.c:1]: Header should be updated in general and a description on the setup required since it requires external hardware

**Review by jmadden173 — APPROVED** [review]
Looks good to merge, there still a few style guide things but its fine for now. In #59 I as looking at adding the linter for the style guide which will make conforming to the style a lot easier.

Before you merge please add your changes to `CHANGELOG.md` and test the system once merged to check for any random errors.

Also it's really helpful in the PR description to add how to test the code for someone reviewing.
